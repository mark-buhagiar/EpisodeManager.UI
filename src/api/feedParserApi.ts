import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import ParseHistoryListing from '../models/ParseHistoryListing';
import ParseHistoryDetails from '../models/ParseHistoryDetails';
import moment from 'moment';

const buildEndpoint = (endpoint: string): string => ApiConfig.feedParserApiBase + endpoint;

const feedParserMSEndpoints = {
    syncWithShowRss: buildEndpoint('sync/SyncWithShowRss'),
    forceSyncParseHistory: buildEndpoint('sync/ForceSyncParseHistory'),
    getParseHistoryBetweenDates: buildEndpoint('ParseHistory/GetParseHistoryBetweenDates'),
    getParseHistoryDetails: buildEndpoint('ParseHistory/GetParseHistoryDetails'),
};

export async function syncWithShowRss(): Promise<void> {
    try {
        await axios.post(feedParserMSEndpoints.syncWithShowRss);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

function convertDates(listing: ParseHistoryListing): ParseHistoryListing {
    return {
        ...listing,
        retrievedOn: moment.utc(listing.retrievedOn).toDate(),
        lastSyncedOn: moment.utc(listing.lastSyncedOn).toDate(),
    };
}

export async function getParseHistoryBetweenDates(startDate: Date, endDate: Date): Promise<ParseHistoryListing[]> {
    try {
        const params = {
            startDate,
            endDate,
        };
        const parseHistory = await axios.get<ParseHistoryListing[]>(feedParserMSEndpoints.getParseHistoryBetweenDates, {
            params,
        });

        return parseHistory.data.map(convertDates);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getParseHistoryDetails(id: number): Promise<ParseHistoryDetails> {
    try {
        const params = {
            parseHistoryId: id,
        };
        const parseHistoryDetails = await axios.get<ParseHistoryDetails>(feedParserMSEndpoints.getParseHistoryDetails, {
            params,
        });

        return parseHistoryDetails.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
export async function forceSyncParseHistory(id: number): Promise<void> {
    try {
        const params = {
            parseHistoryId: id,
        };

        await axios.post(feedParserMSEndpoints.forceSyncParseHistory, null, {
            params,
        });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
