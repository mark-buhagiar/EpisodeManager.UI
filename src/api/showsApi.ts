import axios from 'axios';
import Episode from '../models/Episode';
import ApiConfig from '../config/apiConfig';

const buildEndpoint = (endpoint: string): string => ApiConfig.showsApiBase + endpoint;

const showsMSEndpoints = {
    getForCurrentUserBetweenDates: buildEndpoint('episodes/getForCurrentUserBetweenDates'),
};

export async function getForCurrentUserBetweenDates(startDate: Date, endDate: Date): Promise<Episode[]> {
    try {
        const params = {
            startDate: startDate,
            endDate: endDate,
        };
        const result = await axios.get<Episode[]>(showsMSEndpoints.getForCurrentUserBetweenDates, { params });
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
