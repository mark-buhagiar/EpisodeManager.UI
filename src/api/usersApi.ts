import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import EpisodePreferences from '../models/EpisodePreferences';
import Show from '../models/Show';

const buildEndpoint = (endpoint: string): string => ApiConfig.usersApiBase + endpoint;

const usersMSEndpoints = {
    addEpisodeDownloaded: buildEndpoint('subscriptions/AddEpisodeDownloaded'),
    getUserSubscriptionsForListing: buildEndpoint('subscriptions/GetUserSubscriptionsForListing'),
    unsubscribeFromShow: buildEndpoint('subscriptions/UnsubscribeFromShow'),
    subscribeToShow: buildEndpoint('subscriptions/SubscribeToShow'),
    setUserEpisodePreferences: buildEndpoint('users/setUserEpisodePreferences'),
    getUserEpisodePreferences: buildEndpoint('users/getUserEpisodePreferences'),
};

export async function addEpisodeDownloaded(episodeId: number): Promise<void> {
    try {
        const params = {
            episodeId,
        };
        await axios.put(usersMSEndpoints.addEpisodeDownloaded, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserSubscriptionsForListing(): Promise<Show[]> {
    try {
        const result = await axios.get(usersMSEndpoints.getUserSubscriptionsForListing);
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function unsubscribeFromShow(showId: number): Promise<void> {
    try {
        const params = {
            showId,
        };
        await axios.patch(usersMSEndpoints.unsubscribeFromShow, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function subscribeToShow(showIds: number[]): Promise<void> {
    try {
        const params = {
            showIds,
        };
        await axios.put(usersMSEndpoints.subscribeToShow, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function setUserEpisodePreferences(preferences: EpisodePreferences): Promise<void> {
    try {
        await axios.post(usersMSEndpoints.setUserEpisodePreferences, preferences);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserEpisodePreferences(): Promise<EpisodePreferences> {
    try {
        const response = await axios.get(usersMSEndpoints.getUserEpisodePreferences);
        return response.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
