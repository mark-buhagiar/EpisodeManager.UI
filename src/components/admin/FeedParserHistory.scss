.parse-history-summary {
    display: flex;
    flex-flow: row nowrap;
    width: 100%;

    * {
        flex-basis: 33%;
    }
}
