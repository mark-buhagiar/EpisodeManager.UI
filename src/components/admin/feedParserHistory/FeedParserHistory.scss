.parse-history-summary {
    display: flex;
    flex-flow: row nowrap;
    width: 100%;

    * {
        flex-basis: 33%;
    }
}

.no-results {
    text-align: center;
}
