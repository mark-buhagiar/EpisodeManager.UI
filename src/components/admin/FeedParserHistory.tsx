import React, { useState } from 'react';
import * as feedParserApi from '../../api/feedParserApi';
import ParseHistoryListing from '../../models/ParseHistoryListing';
import DateRange from '../common/DateRange';
import Panel from '../common/panel';
import FeedParserHistoryItem from './FeedParserHistoryItem';
import './FeedParserHistory.scss';

const FeedParserHistory: React.FC = (): JSX.Element => {
    const today = new Date();
    const [parseHistoryItems, setParseHistoryItems] = useState<ParseHistoryListing[]>([]);

    async function handleDateRangeChanged(startDate: Date, endDate: Date): Promise<void> {
        const feedParserHistory = await feedParserApi.getParseHistoryBetweenDates(startDate, endDate);
        setParseHistoryItems(feedParserHistory);
    }

    const panelActions = (
        <DateRange initialStartDate={today} initialEndDate={today} onDateRangeChanged={handleDateRangeChanged} />
    );

    return (
        <Panel title="Feed Parser History" panelActions={panelActions}>
            {parseHistoryItems
                .sort((a, b) => b.lastSyncedOn.getTime() - a.lastSyncedOn.getTime())
                .map((parseHistoryItem) => (
                    <FeedParserHistoryItem key={parseHistoryItem.id} parseHistoryItem={parseHistoryItem} />
                ))}
        </Panel>
    );
};

export default FeedParserHistory;
