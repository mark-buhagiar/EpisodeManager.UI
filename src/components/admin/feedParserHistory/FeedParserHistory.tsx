import React, { useState } from 'react';
import * as feedParserApi from '../../../api/feedParserApi';
import ParseHistoryListing from '../../../models/ParseHistoryListing';
import DateRange from '../../common/DateRange';
import Panel from '../../common/panel';
import ComponentLoading from '../../loading/ComponentLoading';
import './FeedParserHistory.scss';
import FeedParserHistoryItem from './FeedParserHistoryItem';
import ComponentError from '../../error/ComponentError';

const FeedParserHistory: React.FC = (): JSX.Element => {
    const today = new Date();
    const [parseHistoryItems, setParseHistoryItems] = useState<ParseHistoryListing[]>([]);
    const [loading, setIsLoading] = useState(false);
    const [hasErrored, setHasErrored] = useState(false);

    async function handleDateRangeChanged(startDate: Date, endDate: Date): Promise<void> {
        setIsLoading(true);
        try {
            const feedParserHistory = await feedParserApi.getParseHistoryBetweenDates(startDate, endDate);
            setParseHistoryItems(feedParserHistory);
        } catch (ex) {
            setHasErrored(true);
        }
        setIsLoading(false);
    }

    const panelActions = (
        <DateRange initialStartDate={today} initialEndDate={today} onDateRangeChanged={handleDateRangeChanged} />
    );

    const body = (): React.ReactNode => {
        if (loading) {
            return <ComponentLoading />;
        } else if (hasErrored) {
            return <ComponentError />;
        } else if (parseHistoryItems.length === 0) {
            return <span className="no-results">No Feed Parser Data Available</span>;
        } else {
            return parseHistoryItems
                .sort(
                    (a: ParseHistoryListing, b: ParseHistoryListing) =>
                        b.lastSyncedOn.getTime() - a.lastSyncedOn.getTime(),
                )
                .map((parseHistoryItem) => (
                    <FeedParserHistoryItem key={parseHistoryItem.id} parseHistoryItem={parseHistoryItem} />
                ));
        }
    };

    return (
        <Panel title="Feed Parser History" panelActions={panelActions}>
            {body()}
        </Panel>
    );
};

export default FeedParserHistory;
