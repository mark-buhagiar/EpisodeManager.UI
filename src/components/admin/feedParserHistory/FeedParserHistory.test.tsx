import { act, cleanup, fireEvent, render } from '@testing-library/react';
import moment from 'moment';
import React from 'react';
import { getParseHistoryBetweenDates } from '../../../api/feedParserApi';
import FeedParserHistory from './FeedParserHistory';

jest.mock('@material-ui/pickers', () => {
    const actual = jest.requireActual('@material-ui/pickers');
    const moment = jest.requireActual('moment');

    return {
        ...actual,
        DatePicker: (() => (props: any): React.ReactNode => {
            return (
                <input
                    data-testid="mockedDateField"
                    onChange={(e): void => {
                        props.onChange(moment(e.target.value));
                    }}
                />
            );
        })(),
    };
});

jest.mock('../../../api/feedParserApi', () => ({ getParseHistoryBetweenDates: jest.fn() }));

const validResult = [
    {
        id: 28,
        retrievedOn: moment('2020-04-13T15:28:27.751938').toDate(),
        lastSyncedOn: moment('2020-04-13T15:28:27.751938').toDate(),
        episodeCount: 100,
        syncCount: 1,
        success: true,
    },
    {
        id: 30,
        retrievedOn: moment('2020-04-16T22:15:03.898613').toDate(),
        lastSyncedOn: moment('2020-04-16T22:15:03.898613').toDate(),
        episodeCount: 100,
        syncCount: 1,
        success: true,
    },
];

afterEach(async () => {
    await cleanup();
    jest.clearAllMocks();
});

describe('When the component is loading', () => {
    it('should display a loading spinner', async (): Promise<void> => {
        (getParseHistoryBetweenDates as jest.Mock).mockReturnValueOnce(Promise.resolve([...validResult]));
        const { asFragment } = render(<FeedParserHistory />);
        expect(asFragment()).toMatchSnapshot();

        await act(async () => {
            await Promise.resolve();
        });
    });
});

describe('When data retrieval errors out', () => {
    it('should display an error icon', async (): Promise<void> => {
        (getParseHistoryBetweenDates as jest.Mock).mockReturnValueOnce(Promise.reject('Error test'));
        const { asFragment } = render(<FeedParserHistory />);

        await act(async () => {
            await Promise.resolve();
        });
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('When data retrieval completes successfully', () => {
    it('should display the appropriate message if there are no results', async (): Promise<void> => {
        (getParseHistoryBetweenDates as jest.Mock).mockReturnValueOnce(Promise.resolve([]));
        const { asFragment, getByText } = render(<FeedParserHistory />);

        await act(async () => {
            await Promise.resolve();
        });
        expect(asFragment()).toMatchSnapshot();
        expect(getByText('No Feed Parser Data Available')).toBeTruthy();
    });

    it('should display a list of results', async (): Promise<void> => {
        (getParseHistoryBetweenDates as jest.Mock).mockReturnValueOnce(Promise.resolve([...validResult]));
        const { asFragment, queryByText, getAllByTestId } = render(<FeedParserHistory />);

        await act(async () => {
            await Promise.resolve();
        });

        expect(asFragment()).toMatchSnapshot();
        expect(queryByText('No Feed Parser Data Available')).toBeNull();

        const sortedResults = [...validResult];
        sortedResults.sort((a, b) => (a.lastSyncedOn.getTime() < b.lastSyncedOn.getTime() ? 1 : -1));

        const commonTestId = 'feed-parser-history-item';
        const historyItems = getAllByTestId(commonTestId, { exact: false });
        expect(historyItems.length).toEqual(validResult.length);

        historyItems.forEach((renderedRow, index) => {
            expect(renderedRow.getAttribute('data-testId')).toEqual(`${commonTestId}-${sortedResults[index].id}`);
        });
    });
});

describe('When selected dates are changed', () => {
    it('should re-query the web service', async (): Promise<void> => {
        (getParseHistoryBetweenDates as jest.Mock).mockResolvedValue([...validResult]);

        await act(async () => {
            const { getAllByTestId } = render(<FeedParserHistory />);
            await Promise.resolve();
            const [startDate, endDate] = getAllByTestId('mockedDateField');
            fireEvent.change(startDate, { target: { value: moment().subtract(2, 'weeks').format() } });
            await Promise.resolve();
            fireEvent.change(endDate, { target: { value: moment().subtract(1, 'weeks').format() } });
            await Promise.resolve();
        });

        expect(getParseHistoryBetweenDates).toHaveBeenCalledTimes(3);
    });
});
