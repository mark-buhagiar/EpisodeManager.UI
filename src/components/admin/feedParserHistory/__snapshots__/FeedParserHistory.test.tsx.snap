// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When data retrieval completes successfully should display a list of results 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Feed Parser History
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <div
      class="MuiPaper-root MuiExpansionPanel-root MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
      data-testid="feed-parser-history-item-30"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiExpansionPanelSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiExpansionPanelSummary-content"
        >
          <div
            class="parse-history-summary"
          >
            <div
              class="date"
              title="Last synced on"
            >
              16 April 2020 22:15
            </div>
            <div
              class="episode-count"
              title="Episode count"
            >
              100
            </div>
            <div
              class="status"
            >
              Success
            </div>
          </div>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <span
              class="material-icons"
            >
              expand_more
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiExpansionPanel-root MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
      data-testid="feed-parser-history-item-28"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiExpansionPanelSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiExpansionPanelSummary-content"
        >
          <div
            class="parse-history-summary"
          >
            <div
              class="date"
              title="Last synced on"
            >
              13 April 2020 15:28
            </div>
            <div
              class="episode-count"
              title="Episode count"
            >
              100
            </div>
            <div
              class="status"
            >
              Success
            </div>
          </div>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <span
              class="material-icons"
            >
              expand_more
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`When data retrieval completes successfully should display the appropriate message if there are no results 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Feed Parser History
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <span
      class="no-results"
    >
      No Feed Parser Data Available
    </span>
  </div>
</DocumentFragment>
`;

exports[`When data retrieval errors out should display an error icon 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Feed Parser History
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <div
      class="component-error-container"
    >
      <span
        class="component-error"
      >
        <span
          class="material-icons"
        >
          error_outline
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`When the component is loading should display a loading spinner 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Feed Parser History
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <div
      class="component-loading-container"
    >
      <span
        class="component-loading"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
